### ISSUE TITLE
hard to get memq using member-with

### ISSUE THREAD

**Issue by shriram** [brownplt/pyret-lang#1223]
`member` uses `==`. Sometimes it's useful to have `member` with the other built-in equality primitives (like `identical` to get "memq"). 

In principle, `member-with` is ideal for this. However, because the third argument for `member-with` is so general, the standard equality predicates don't fit in it, and it's awfully convoluted to convert the built-ins to actually work with this framework — I tried it and nearly got hopelessly lost.

Can we have an easier way of adapting things like `member` to be able to just use the other built-in primitives? Just assume the equality predicate returns true/false and make it easy to provide that as a third argument?

**Comment by jpolitz** [comment]
Hm, yeah. The easiest way is

```
import to-boolean from equality

to-boolean(lists.member-with([list: 1, 2, 3], 3, identical3))
```

Which is silly. This function should really be called `member-with3`, and `member-with` should take a boolean-producing function.

**Comment by blerner** [comment]
Eh, not 100% sure I want that -- suppose I had a [list: some-non-matching-value, some-not-comparable-value, some-matching-value].  Do I want the member-with call to blow up on the middle value, when it's possible a later value might work?

**Comment by shriram** [comment]
But Ben, the convention we've adopted (if I'm not mistaken) _is_ that equality-things with `3` at the end are three-valued and hence never blow up, while those that aren't are `Boolean` and may blow up.

**Comment by blerner** [comment]
Yes, "may blow up" ... if necessary.  I.e. I can write `(5 == lam(x): 3 end) is false` without erroring, despite the fact that functions aren't comparable, because one is-Number and one is-Function.  My question was probing whether we think this sort of member-with function constitutes a necessary situation.  If we decide that it should be, ok.  I just want to make it a deliberate decision to do so, not an inadvertent consequence of a particular implementation of member-with.

**Comment by jpolitz** [comment]
The proposal is that `member(l, elt)` should mean `member-with(l, elt, equal-always)`

Currently if you search a list of roughnums for a number using `member`, it's an error and that's fine. I think the same should be true of `member-with`.

If you _really_ need to do potentially Unknown comparisons, use `member-with3`

**Comment by blerner** [comment]
Gotcha, ok.  I was thinking of a higher-order case, e.g. of a list of objects-containing-functions, and thinking through what kinds of crashy weirdness might happen there.  But I'm ok with this as a design decision; I just wanted it to be clarified.

**Comment by shriram** [comment]
Great. I think we're converged then. 

It would be helpful if a bit of this rationale were in the docs, in particular _with a link to the `3` versions_, since to a first-time Pyreteer there's nothing obvious about the existence of the latter, and without the docs saying "you might want to look over there, too", they most probably wouldn't. Essentially, we need instead of just 2.3.8 ("Total Equality Functions (Avoiding Incomparability Errors)"), "Fallible Equality" and "Infallible Equality" pointing to each other.
